src/adapters/fauna/index.js

```diff
@@ -0,0 +1,420 @@
+// create faunadb server key
+// create collections: users, accounts, sessions, verificationRequests
+// create required indexes
+
+import faunadb, { query as q } from 'faunadb'
+import { v4 as uuidv4 } from 'uuid'
+import { createHash, randomBytes } from 'crypto'
+
+const INDEX_USERS_ID = 'index_users_id'
+const INDEX_USERS_EMAIL = 'index_users_email'
+const INDEX_USERS_ACCOUNT_ID_PROVIDER_ID = 'index_accounts_providerId_providerAccountId'
+const INDEX_VERIFICATION_REQUESTS_TOKEN = 'index_verificationRequests_token'
+const INDEX_SESSIONS_ID = 'index_sessions_id'
+const INDEX_SESSIONS_SESSION_TOKEN = 'index_sessions_sessionToken'
+
+const serverClient = new faunadb.Client({ secret: process.env.FAUNADB_SERVER_KEY })
+
+function faunaWrapper (faunaQ, errorTag) {
+  try {
+    return serverClient.query(faunaQ)
+  } catch (error) {
+    console.error(errorTag, error)
+    return Promise.reject(new Error(errorTag, error))
+  }
+}
+
+const Adapter = (config, options = {}) => {
+  async function getAdapter (appOptions) {
+    if (appOptions && (!appOptions.session || !appOptions.session.maxAge)) {
+      console.log('no default options for session')
+    }
+
+    const defaultSessionMaxAge = 30 * 24 * 60 * 60 * 1000
+    const sessionMaxAge =
+            appOptions && appOptions.session && appOptions.session.maxAge
+              ? appOptions.session.maxAge * 1000
+              : defaultSessionMaxAge
+    const sessionUpdateAge =
+            appOptions && appOptions.session && appOptions.session.updateAge
+              ? appOptions.session.updateAge * 1000
+              : 24 * 60 * 60 * 1000
+
+    async function createUser (profile) {
+      console.log('-----------createUser------------')
+      // console.log(profile);
+      return faunaWrapper(
```

Comment:
I'm not sure if this is a pattern with fauna (new user as well) but to use the ref's across the other methods I basically return:
```
return await faunaClient.query(q.Create(q.Collection('users'), { data: profile }))
          .then(res => { return { ...res.data, id: res.ref } })
```

This way further down on other methods like getUser, I can directly use the id as follows:
```
    async function getUser (id) {
      try {
        return await faunaClient
          .query(q.Get(q.Ref(id)))
          .then(res => { return { ...res.data, id: res.ref } })
          .catch(err => Promise.resolve(null))
      } catch (error) {
        return Promise.reject(new Error('GET_USER_ERROR'), error)
      }
    }
```

Reply:
It is a fauna pattern. It's optimal to use FQL way to execute multiple operations instead of using js patterns.

Reply:
maybe it wasn't clear because of the line # of the comment, but I wasn't referring to the faunaWrapper. I was referring to the `q.Select('data'` part ... by returning the ref object as well, I was able to use the fauna IDs directly (Ref) rather than creating yet another set of `id:` fields for each object. 

Which I'm not sure if fauna encourages that or encourages devs to handle their own IDs.

Reply:
You are correct. We can directly use fauna ID but I just want to have id field with uuid so that I can simply extract the 'data' part of the document in case of migration to other db.